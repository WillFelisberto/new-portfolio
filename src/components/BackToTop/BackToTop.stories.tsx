import { Meta, StoryObj } from '@storybook/react';
import BackToTop from '.';

const meta: Meta<typeof BackToTop> = {
  title: 'Components/BackToTop',
  component: BackToTop,
  parameters: {
    layout: 'fullscreen',
  },
};

export default meta;

type Story = StoryObj<typeof BackToTop>;

export const Default: Story = {
  render: () => (
    <div style={{ height: '2000px', padding: '1rem' }}>
      <p>Scroll down to see the BackToTop button.</p>
      <BackToTop />
    </div>
  ),
};
