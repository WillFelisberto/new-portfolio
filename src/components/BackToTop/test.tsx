import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import BackToTop from '.';

describe('BackToTop component', () => {
  beforeEach(() => {
    // Mockando o scrollY e o scrollTo
    Object.defineProperty(window, 'scrollY', {
      writable: true,
      configurable: true,
      value: 0,
    });

    Object.defineProperty(window, 'scrollTo', {
      writable: true,
      value: jest.fn(),
    });
  });

  const triggerScroll = (scrollValue: number) => {
    Object.defineProperty(window, 'scrollY', { value: scrollValue });
    fireEvent.scroll(window); // Dispara o evento de rolagem
  };

  it('should render the BackToTop component', () => {
    const { container } = render(<BackToTop />);
    triggerScroll(600); // Simula um scrollY > 500

    expect(
      screen.getByRole('button', { name: /back to top/i }),
    ).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('does not render the button when scrollY is less than 500', () => {
    render(<BackToTop />);
    expect(
      screen.queryByRole('button', { name: /back to top/i }),
    ).not.toBeInTheDocument();
  });

  it('renders the button when scrollY is greater than 500', () => {
    render(<BackToTop />);
    triggerScroll(600); // Simula um scrollY > 500
    expect(
      screen.getByRole('button', { name: /back to top/i }),
    ).toBeInTheDocument();
  });

  it('hides the button when scrollY is less than or equal to 500 after being shown', async () => {
    render(<BackToTop />);
    triggerScroll(600); // Mostra o botão

    // Aguarda a exibição do botão
    await waitFor(() => {
      expect(
        screen.getByRole('button', { name: /back to top/i }),
      ).toBeInTheDocument();
    });

    triggerScroll(400); // Esconde o botão

    // Aguarda a remoção do botão
    await waitFor(() => {
      expect(
        screen.queryByRole('button', { name: /back to top/i }),
      ).not.toBeInTheDocument();
    });
  });

  it('scrolls to the top when the button is clicked', () => {
    render(<BackToTop />);
    triggerScroll(600); // Mostra o botão
    const button = screen.getByRole('button', { name: /back to top/i });
    fireEvent.click(button);
    expect(window.scrollTo).toHaveBeenCalledWith({
      top: 0,
      behavior: 'smooth',
    });
  });
});
