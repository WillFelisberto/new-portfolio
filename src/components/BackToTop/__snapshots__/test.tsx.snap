// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackToTop component should render the BackToTop component 1`] = `
<div
  class="fixed bottom-4 right-4 z-20"
  style="opacity: 0; right: -10px;"
>
  <button
    aria-label="Back to top"
    class="flex items-center justify-center gap-2 rounded-lg bg-sky-600 px-4 py-3 text-gray-50 transition-all hover:bg-sky-500 disabled:opacity-50 shadow-lg shadow-sky-400/20"
  >
    <svg
      fill="none"
      height="20"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 5l0 14"
      />
      <path
        d="M16 9l-4 -4"
      />
      <path
        d="M8 9l4 -4"
      />
    </svg>
  </button>
</div>
`;
